d2/src/main/java/com/linkedin/d2/balancer/servers/ZooKeeperAnnouncer.java

```diff
@@ -833,4 +854,59 @@ private ImmutablePair<String, String> getZnodePathAndData(String cluster) {
   public boolean isWarmingUp() {
     return _isWarmingUp;
   }
+
+  private Map<Integer, PartitionData> validatePartitionData(Map<Integer, PartitionData> partitionData) {
+    if (_maxWeight == null || _actOnWeightBreach == ActOnWeightBreach.IGNORE) {
+      return partitionData;
+    }
+
+    Map<Integer, PartitionData> res = new HashMap<>(partitionData); // modifiable copy in case the input is unmodifiable
+    for (Map.Entry<Integer, PartitionData> entry : res.entrySet()) {
+      BigDecimal weight = BigDecimal.valueOf(entry.getValue().getWeight());
+      // check max weight
+      if (weight.compareTo(_maxWeight) > 0) {
+        switch (_actOnWeightBreach) {
+          case WARN:
+            _log.warn("", getMaxWeightBreachException(weight, entry.getKey()));
+            break;
+          case THROW:
+            throw getMaxWeightBreachException(weight, entry.getKey());
+          case RECTIFY:
+            entry.setValue(new PartitionData(_maxWeight.intValue()));
+            _log.warn("Capped weight {} in Partition {} to the max weight allowed: {}.", weight.doubleValue(),
+                entry.getKey(), _maxWeight.intValue());
+            break;
+          default:
+            break;
+        }
+      }
+
+      // check decimal places
+      if (weight.scale() > _maxWeight.scale()) {
+        switch (_actOnWeightBreach) {
+          case WARN: // both WARN and THROW only log the warning. Don't throw exception for decimal places.
+          case THROW:
+            _log.warn("", new IllegalArgumentException(String.format("Weight %f in Partition %d has more than %d"
+                + " decimal places. It will be rounded in the future.", weight.doubleValue(), entry.getKey(),
+                _maxWeight.scale())));
+            break;
+          case RECTIFY:
+            double newWeight = weight.setScale(_maxWeight.scale(), RoundingMode.HALF_UP).doubleValue();
+            entry.setValue(new PartitionData(newWeight));
```

Comment:
I think this will override the `entry.setValue` in the previous `switch` if the provided weight is both bigger than the max weight and has more decimal places